plugins/gcp/resoto_plugin_gcp/resources/compute.py

```diff
@@ -2714,13 +2720,49 @@ class GcpInstance(GcpResource, BaseInstance):
     instance_status_message: Optional[str] = field(default=None)
     instance_tags: Optional[GcpTags] = field(default=None)
 
+    def set_machine_type(self, machine_type_link_ish: str, graph_builder: GraphBuilder) -> None:
+        def find_existing_machine_type() -> Optional[GcpMachineType]:
+            if not machine_type_link_ish.startswith("https://"):
+                machine_type_link_ish = (
+                    f"https://www.googleapis.com/compute/v1/projects/{graph_builder.project.id}/{machine_type_link_ish}"
+                )
+            return graph_builder.node(clazz=GcpMachineType, link=machine_type_link_ish)
+
+        # TODO: Maybe as class method on GcpMachineType?
+        def create_adhoc_machine_type() -> GcpMachineType:
+            log.debug(f"Creating ad-hoc machine type for {self.rtdname}")
+            zone = machine_type_link_ish.split("/")[-3]
+            name = machine_type_link_ish.split("/")[-1]
+            result = graph_builder.client.get(
+                gcp_service="compute",
+                action="get",
+                resource="machineType",
+                project=graph_builder.project,
+                zone=zone,
+                machineType=name,
+            )
+            return GcpMachineType.from_api(result)
+
+        machine_type = find_existing_machine_type() or create_adhoc_machine_type()
```

Comment:
The current collector detects a custom machine type by this line `if resource.instance_type in ("", None) and "custom" in resource._machine_type_link`.
Plus: you need to make sure to not create multiple instances for the same custom machine type. 
After collecting all machine instances, you know all custom machine types.

Reply:
>  Plus: you need to make sure to not create multiple instances for the same custom machine type.

Sorry, what do you mean? We are relying on the "if not found then create adhoc", the create-part of which would only happen for the very first `GcpInstance` of each (project, zonal) `GcpMachineType`.

Reply:
Making requests to GCP happens asynchronously, while the list of resources is walked synchronously.
Waling 2 instances with the same custom instance type could yield fetching the same instance type twice.